tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+Respectively, you have to carry expected responses or even (rudimentary) API behavior in your code base.
+You now have two relationships to keep in mind: your code must pass your mocked/faked tests, and your mocks/fakes must correspond to the live API.
+
+This tradeoff cannot be resolved, but it is eased considerably by decoupled (or modular) code.
+Much as with other side effects, it can help to concentrate your API calls in as few functions as possible and test these with mocks or fakes, where necessary.
+All other functions can then be tested without dependence on the API.
+Conversely, if you find yourself requiring these test doubles in a lot of places, you may want to [consider a refactoring](https://medium.com/javascript-scene/mocking-is-a-code-smell-944a70c90a6a).
+
+## Unit vs Integration Testing
+
+Test doubles also benefit from a clean separation of concerns.
+
+On the one hand, [*unit tests*](https://en.wikipedia.org/wiki/Unit_testing) should only test the behavior of your code, not the behavior of the API.
+Otherwise a failed test could be ambiguous: Was it your code, or a errant API behavior?
+Unit tests will typically cover simple, but user facing behavior, such as whether an API response is parsed correctly, or whether an HTTP error code is handled appropriately.
+Mocks, especially, can help you write isolated unit tests.
+
+On the other hand, [*integration tests*](https://en.wikipedia.org/wiki/Integration_testing) should test how your code interacts with other software or services.
+["Narrow" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against test doubles, typically fakes.
+This may be most helpful to test more complex, perhaps stateful, interactions between your code and the API, which relies on internal logic of the API rather than a one-off response.
+For example, you may want to test whether if you submit a string to a translation service, you get back a result in the appropriate language.
+Such test duplication of external business logic may be hard to get right and maintain, and should rarely be necessary.
+
+["Broad" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against live services, and thus require no test doubles.
+
+Neither unit nor integration tests need to cover an API specification comprehensively.
+Testing the API is not your concern, but merely the narrow interface between the API and your code.
+As a rule of thumb, your code should handle, and your tests should cover the failure *modes* of the API, including HTTP response codes, API response codes or incorrect response formats.
+Perhaps you can additionally test some API requests against some known outputs, but this will always remain a sample.
+There's always a chance that some deeply nested JSON response will break your code.
+You may not be able to cover all such errors from an errand or changed API, and that's okay.
+
+## Example
```

Comment:
I like seeing an example and wonder how it could be more coupled with the theoretical concepts from earlier in the chapter. I.e. having the example first, with notions explained with it.

Reply:
My idea was to focus:

1) non-example part on *why* you'd want to minimise your API exposure
2) how to minimise that in a toy example

Might also work well if it is restructured as you suggest, but I don't think I can do that now -- maybe first resolve overall question of tone and direction.